core/commands/pin.go

```diff
@@ -200,25 +212,78 @@ Example:
 			res.SetError(err, cmds.ErrNormal)
 			return
 		}
-		if !found {
-			typeStr = "recursive"
-		}
 
-		switch typeStr {
-		case "all", "direct", "indirect", "recursive":
-		default:
-			err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
-			res.SetError(err, cmds.ErrClient)
+		if found {
+			switch typeStr {
+			case "all", "direct", "indirect", "recursive":
+			default:
+				err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
+				res.SetError(err, cmds.ErrClient)
+			}
 		}
 
 		keys := make(map[string]RefKeyObject)
-		if typeStr == "direct" || typeStr == "all" {
-			for _, k := range n.Pinning.DirectKeys() {
+
+		if len(req.Arguments()) > 0 {
+			if !found {
+				typeStr = "all"
+			}
+
+			for _, p := range req.Arguments() {
+				dagNode, err := core.Resolve(req.Context(), n, path.Path(p))
```

Comment:
(this will be able to use the func in https://github.com/ipfs/go-ipfs/pull/2201/files once it merges.)